anvil/server/src/ipc.rs

```diff
@@ -0,0 +1,178 @@
+    pub fn incoming(self) -> io::Result<impl Stream<Item = impl Future<Output = ()> + Unpin>> {
+        let IpcEndpoint { handler, endpoint } = self;
+        trace!(target: "ipc",  endpoint=?endpoint.path(), "starting ipc server" );
+
+        if cfg!(unix) {
+            // ensure the file does not exist
+            if std::fs::remove_file(endpoint.path()).is_ok() {
+                warn!(target: "ipc", endpoint=?endpoint.path(), "removed existing file");
+            }
+        }
+
+        let connections = match endpoint.incoming() {
+            Ok(connections) => connections,
+            Err(err) => {
+                error!(target: "ipc",  ?err, "Failed to create ipc listener");
+                return Err(err)
+            }
+        };
+
+        trace!(target: "ipc", "established connection listener");
+
+        let connections = connections.filter_map(move |stream| {
+            let handler = handler.clone();
+            Box::pin(async move {
+                if let Ok(stream) = stream {
+                    trace!(target: "ipc", "successful incoming IPC connection");
+                    let framed = tokio_util::codec::Decoder::framed(JsonRpcCodec, stream);
+                    Some(PubSubConnection::new(IpcConn(framed), handler))
+                } else {
+                    None
+                }
+            })
+        });
+        Ok(connections)
+    }
+}
+
+#[pin_project::pin_project]
+struct IpcConn<T>(#[pin] T);
+
+impl<T> Stream for IpcConn<T>
+where
+    T: Stream<Item = io::Result<String>>,
+{
+    type Item = Result<Option<Request>, RequestError>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        fn on_request(msg: io::Result<String>) -> Result<Option<Request>, RequestError> {
+            let text = msg?;
+            Ok(Some(serde_json::from_str(&text)?))
+        }
+        match ready!(self.project().0.poll_next(cx)) {
+            Some(req) => Poll::Ready(Some(on_request(req))),
+            _ => Poll::Ready(None),
+        }
+    }
+}
+
+impl<T> Sink<String> for IpcConn<T>
+where
+    T: Sink<String, Error = io::Error>,
+{
+    type Error = io::Error;
+
+    fn poll_ready(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // NOTE: we always flush here this prevents any backpressure buffer in the underlying
+        // `Framed` impl that would cause stalled requests
+        self.project().0.poll_flush(cx)
+    }
+
+    fn start_send(self: Pin<&mut Self>, item: String) -> Result<(), Self::Error> {
+        self.project().0.start_send(item)
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.project().0.poll_flush(cx)
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.project().0.poll_close(cx)
+    }
+}
+
+struct JsonRpcCodec;
```

Comment:
Are we OK with writing our own codecs? Is this something we should start refactoring, eg for have in reth?

Reply:
the current state of the different codec traits and AsyncReadWrite + Stream/Sink is a bit of a mess...

we probably only need this when we something that's only AsyncWrite/AsyncRead but we need a stream,

The way this currently works is `Framed::new(asyncread+asyncwrite, codec)` the codec here determines when a new `framed` object can be produced as `Stream::Item`
 
this codec basically finds enclosed json objects via `{` and determines when a full object is in the buffer and can be deserialized.